httpbp/response.go

```diff
@@ -0,0 +1,135 @@
+package httpbp
+
+import (
+	"context"
+	"encoding/json"
+	"html/template"
+	"net/http"
+
+	httpgk "github.com/go-kit/kit/transport/http"
+)
+
+// ResponseCookies is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically add cookies to the
+// response.
+type ResponseCookies interface {
+	// Return a list of all cookies to set on the response.
+	Cookies() []*http.Cookie
+
+	// Add a new cookie to set on the response.
+	AddCookie(*http.Cookie)
+}
+
+// ResponseHeaders is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically add http headers to the
+// response.
+type ResponseHeaders interface {
+	// Headers returns the http.Header container that can be used to add headers
+	// to the response.
+	Headers() http.Header
+}
+
+// ErrorResponse is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically return http errors.
+type ErrorResponse interface {
+	// Err returns the error set on the response.
+	//
+	// This method is named "Err" rather than "Error" to avoid confusion with
+	// the "Error" function that is specified in the error interface.
+	Err() error
+
+	// SetError sets an error to be returned in the response.
+	SetError(err error)
+}
+
+// ErrorMapper is a function used by the httpbp.Encode functions to map the
+// errors provided by responses that implement the ErrorResponse interface to
+// HTTP errors.
+type ErrorMapper func(err error) error
+
+// DefaultErrorMapper is the ErrorMapper used by the httpbp.Encode functions
+// that do not take an ErrorMapper as an argument.
+func DefaultErrorMapper(err error) error {
+	return err
+}
+
+type responseEncoder func(w http.ResponseWriter, r interface{}) error
+
+type encodeArgs struct {
+	contentType string
+	encoder     responseEncoder
+	errorMapper *ErrorMapper
+}
+
+func encodeResponse(w http.ResponseWriter, response interface{}, args encodeArgs) error {
+	if resp, ok := response.(ResponseHeaders); ok {
+		w.Header().Set("Content-Type", args.contentType)
+		for key, values := range resp.Headers() {
+			for _, value := range values {
+				w.Header().Add(key, value)
+			}
+		}
+	}
+
+	if resp, ok := response.(ResponseCookies); ok {
+		for _, cookie := range resp.Cookies() {
+			http.SetCookie(w, cookie)
+		}
+	}
+
+	if resp, ok := response.(ErrorResponse); ok {
+		var mapper ErrorMapper
+		if args.errorMapper == nil {
+			mapper = DefaultErrorMapper
+		} else {
+			mapper = *args.errorMapper
+		}
+		if resp.Err() != nil {
+			return mapper(resp.Err())
+		}
```

Comment:
this can be simplified as (after removed the pointer to `errorMapper`):

```
if args.errorMapper == nil {
  args.errorMapper = DefaultErrorMapper
}
if resp.Err() != nil {
  return args.errorMapper(resp.Err())
}
```